### DISCUSSION TITLE
Transiting to abortable error state

### DISCUSSION THREAD

**Discussion by MohTaheri** [spring-projects/spring-kafka#2789] [Q&A]
Hello

My spring application setup involves three instances where the Kafka producers and consumers are within the same service. i'm using the **`@Transactional`** annotation for consumers with a read_committed isolation level (read message + insert into DB + publish output). We are encountering a **CommitFailedException** when re-deploying to Kubernetes. 
However, with each deployment in Kubernetes, after creating new pods and before terminating the old ones, we encounter these errors. 

```
[Producer clientId=a-5489cf557d-gjpvk-producer-17, transactionalId=tx-31afb89f-b79f-4a17-bf89-e0300ef7dfe716] 
Transiting to abortable error state due to org.apache.kafka.clients.consumer.CommitFailedException: 
Transaction offset Commit failed due to consumer group metadata mismatch: Specified group generation id is not valid. |  
-- | --
 
 thread=[org.springframework.kafka.KafkaListenerEndpointContainer#10-1-C-1] message=Transaction rolled back  org.apache.kafka.clients.consumer.CommitFailedException: 
 Transaction offset Commit failed due to consumer group metadata mismatch: Specified group generation id is not valid. 	
 at org.apache.kafka.clients.producer.internals.TransactionManager$TxnOffsetCommitHandler.handleResponse(TransactionManager.java:1656) 	at 
 org.apache.kafka.clients.producer.internals.TransactionManager$TxnRequestHandler.onComplete(TransactionManager.java:1207) 	
 at org.apache.kafka.clients.ClientResponse.onComplete(ClientResponse.java:109) 	at 
 org.apache.kafka.clients.NetworkClient.completeResponses(NetworkClient.java:583) 	at 
 org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:575) 	at 
 org.apache.kafka.clients.producer.internals.Sender.maybeSendAndPollTransactionalRequest(Sender.java:426) 	at 
 org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:316) 	at 
 org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:243) 	at java.base/java.lang.Thread.run(Unknown Source)

```

It seems this could be related to the `transaction-id-prefix` configuration, causing different transaction IDs for old and new instances. 
Should we consider removing the randomness from this configuration?

```
spring:
  kafka:
    producer:
      transaction-id-prefix: tx-${random.uuid}
```

- spring-kafka: 3.0.10

Thanks

**Comment by garyrussell** [comment]
I think, if you do that, you will just get a different error (`ProducerFencedException`) when the old producer tries to commit. It's best to stop the old producers first to avoid any such errors.

**Reply by garyrussell** [reply]
It's best to reply to my answer rather than adding a new answer below.

You could try setting the `stopImmediate` container property to shut down the old containers faster:

https://github.com/spring-projects/spring-kafka/blob/30ee20da0bb1a1ba39f87f9b5e805cd269cabce1/spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerProperties.java#L911-L920

**Reply by MohTaheri** [reply]
Sorry for that!
Just to clarify, I'm currently calling the `stop()` function from **KafkaListenerEndpointRegistry**, which I believe might have a similar behavior. 
As a result, we didn't face any issues with the first pod. However, the problem still persists with the other pods, which receive the `SIG_TERM` after 20 seconds. In this scenario, is there any option available to stop the other old producers? For instance, could we use the `Git SHA ` instead of `random.uuid` as the **transaction_id_prefix**?

**Reply by garyrussell** [reply]
It's not clear what you mean; producers must have unique ids across instances. I am not sure where the "20 seconds" comes from.

**Reply by MohTaheri** [reply]
Due to the rollingUpdate rules in Kubernetes, all pods won't be deployed simultaneously; they are created one by one. 
 
 As a result, we can't stop all old producers simultaneously

**Reply by garyrussell** [reply]
You could implement your own mechanism - add a "control" listener (with the listener for each instance in its own group) and send a shutdown command to the control topic; that, together with `stopImmediate`, should stop all instances nearly simultaneously. To avoid seeing "old" shutdown commands; use unique groups (e.g. UUID) and `auto.offset.reset=latest`.

**Reply by garyrussell** [reply]
Or, another option would be to set `autoStrartup` to false; then schedule the starts of the new containers to not occur until time has passed for the old instances to stop.

**Reply by MohTaheri** [reply]
Hmm, I'm inclined towards the first solution as well. 
so, it seems that any changes to the `transaction-id-prefix `wouldn't necessarily be helpful for our situation, right?

**Reply by garyrussell** [reply]
I don't believe so, no.

**Reply by MohTaheri** [reply]
ok thanks
So this [answer](https://stackoverflow.com/questions/57179169/how-to-choose-kafka-transaction-id-for-several-applications-hosted-in-kubernete) from you is not valid now.

**Reply by garyrussell** [reply]
That is four years old; KIP-447 changed all that https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics

**Reply by garyrussell** [reply]
I added a note to that answer.

**Reply by MohTaheri** [reply]
None of the solutions can work for us because stopping all previous pods causes a delay in our processing, and we don't want to lose our real-time availability. 
It's strange that exceptions happened for both the producer and consumer sides in both previous and new instances.


```
ERROR thread=[org.springframework.kafka.KafkaListenerEndpointContainer#10-0-C-1] logger=o.s.k.l.KafkaMessageListenerContainer trace_id= span_id= message=Transaction rolled back

org.apache.kafka.clients.consumer.CommitFailedException: Transaction offset Commit failed due to consumer group metadata mismatch: Specified group generation id is not valid.
	at org.apache.kafka.clients.producer.internals.TransactionManager$TxnOffsetCommitHandler.handleResponse(TransactionManager.java:1656)
	at org.apache.kafka.clients.producer.internals.TransactionManager$TxnRequestHandler.onComplete(TransactionManager.java:1207)
	at org.apache.kafka.clients.ClientResponse.onComplete(ClientResponse.java:109)
	at org.apache.kafka.clients.NetworkClient.completeResponses(NetworkClient.java:583)
	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:575)
	at org.apache.kafka.clients.producer.internals.Sender.maybeSendAndPollTransactionalRequest(Sender.java:426)
	at org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:316)
	at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:243)
```



```
ERROR thread=[kafka-producer-network-thread | x-595d98c54b-k5f52-producer-22] logger=o.s.k.s.LoggingProducerListener trace_id= span_id= message=Exception thrown when sending a message with key='078622be-c0f2-4e89-adb8-6e6eae5526ea' and payload='{"is_i...' to topic x:

org.apache.kafka.clients.consumer.CommitFailedException: Transaction offset Commit failed due to consumer group metadata mismatch: Specified group generation id is not valid.
	at org.apache.kafka.clients.producer.internals.TransactionManager$TxnOffsetCommitHandler.handleResponse(TransactionManager.java:1656)
	at org.apache.kafka.clients.producer.internals.TransactionManager$TxnRequestHandler.onComplete(TransactionManager.java:1207)
	at org.apache.kafka.clients.ClientResponse.onComplete(ClientResponse.java:109)
	at org.apache.kafka.clients.NetworkClient.completeResponses(NetworkClient.java:583)
	at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:575)
	at org.apache.kafka.clients.producer.internals.Sender.maybeSendAndPollTransactionalRequest(Sender.java:426)
	at org.apache.kafka.clients.producer.internals.Sender.runOnce(Sender.java:316)
	at org.apache.kafka.clients.producer.internals.Sender.run(Sender.java:243)
	at java.base/java.lang.Thread.run(Unknown Source)
```

Another point that might be related is that we have some recurring consumers producing messages to the topics that are also consumed by another consumer in the same service.

**Reply by garyrussell** [reply]
There is nothing that Spring can do to help with this; the behavior is due to the underlying `kafka-clients` library and the broker itself. I suggest you raise your concerns over there; either via their JIRA or on one of their mailing lists.
